plugins/chat/app/services/chat/list_channel_messages.rb

```diff
@@ -0,0 +1,131 @@
+    step :update_membership_last_viewed_at
+
+    class Contract
+      attribute :channel_id, :integer
+      validates :channel_id, presence: true
+
+      attribute :page_size, :integer
+      validates :page_size,
+                numericality: {
+                  less_than_or_equal_to: ::Chat::MessagesQuery::MAX_PAGE_SIZE,
+                  only_integer: true,
+                },
+                allow_nil: true
+
+      # If this is not present, then we just fetch messages with page_size
+      # and direction.
+      attribute :target_message_id, :integer # (optional)
+      attribute :direction, :string # (optional)
+      attribute :fetch_from_last_read, :boolean # (optional)
+      attribute :target_date, :string # (optional)
+
+      validates :direction,
+                inclusion: {
+                  in: Chat::MessagesQuery::VALID_DIRECTIONS,
+                },
+                allow_nil: true
+    end
+
+    private
+
+    def fetch_channel(contract:, **)
+      ::Chat::Channel.strict_loading.includes(:chatable).find_by(id: contract.channel_id)
+    end
+
+    def fetch_eventual_membership(channel:, guardian:, **)
+      context.membership = channel.membership_for(guardian.user)
+    end
+
+    def enabled_threads?(channel:, **)
+      context.enabled_threads =
+        SiteSetting.enable_experimental_chat_threaded_discussions && channel.threading_enabled
+    end
+
+    def can_view_channel(guardian:, channel:, **)
+      guardian.can_preview_chat_channel?(channel)
+    end
+
+    def determine_target_message_id(contract:, **)
+      contract.target_message_id =
+        context.membership&.last_read_message_id if contract.fetch_from_last_read
+    end
+
+    def target_message_exists(contract:, channel:, guardian:, **)
+      return true if contract.target_message_id.blank?
+      target_message =
+        Chat::Message.with_deleted.find_by(id: contract.target_message_id, chat_channel: channel)
+      return false if target_message.blank?
+      return true if !target_message.trashed?
+      target_message.user_id == guardian.user.id || guardian.is_staff?
+    end
+
+    def fetch_messages(channel:, guardian:, contract:, **)
+      messages_data =
+        ::Chat::MessagesQuery.call(
+          channel: channel,
+          guardian: guardian,
+          target_message_id: contract.target_message_id,
+          page_size: contract.page_size || Chat::MessagesQuery::MAX_PAGE_SIZE,
+          direction: contract.direction,
+          target_date: contract.target_date,
+        )
+
+      context.can_load_more_past = messages_data[:can_load_more_past]
+      context.can_load_more_future = messages_data[:can_load_more_future]
+
+      context.messages = [
+        messages_data[:messages],
+        messages_data[:past_messages]&.reverse,
+        messages_data[:target_message],
+        messages_data[:future_messages],
+      ].flatten.compact
+    end
+
+    def fetch_tracking(guardian:, enabled_threads:, messages:, **)
```

Comment:
Rereading what I had in the old one I had this comment:

```ruby
# Only thread tracking is necessary to fetch here -- we preload
# channel tracking state for all the current user's tracked channels
# in the CurrentUserSerializer.
```

Maybe a better path is to just rename this `fetch_thread_tracking` and `context.thread_tracking` so we don't need the comment to explain why this is not getting channel tracking too.